### PR TITLE
port: Fix double evaluation and filter sensitive settings

### LINKED ISSUE #3735: port: Fix double evaluation and filter sensitive settings (#5641)
The changes in [Fix double evaluation and filter sensitive settings (#5641)](https://github.com/microsoft/botbuilder-dotnet/pull/5641) may need to be ported to maintain parity with `microsoft/botbuilder-dotnet`.

<blockquote>
Fixes #5609

## Description
Fix the double evaluation and filter sensitive settings
</blockquote>

Please review and, if necessary, port the changes.

### PR DESCRIPTION
**Pull request by Danieladu**
Fixes #3735

## Description
Fix the double evaluation and filter sensitive settings

### CODE REVIEW

**Review by boydc2014 — APPROVED** [review]

File: libraries/botbuilder-dialogs/src/memory/scopes/settingsMemoryScope.ts [hunk 1]
```diff
@@ -205,4 +221,29 @@ export class SettingsMemoryScope extends MemoryScope {
...
+    }
+
+    private static deletePropertyPath(obj, path: string): void {
+        if (!obj || !path) {
+            return;
+        }
+        const pathArray = path.split(':');
+
+        for (let i = 0; i < pathArray.length - 1; i++) {
+            const realKey = Object.keys(obj).find((key) => key.toLowerCase() === pathArray[i].toLowerCase());
```
Comment (joshgummersall) [inline libraries/botbuilder-dialogs/src/memory/scopes/settingsMemoryScope.ts:238]: It's probably worth building an index of lowercase key to resolved key once so you aren't enumerating all the keys multiple times.

```typescript
const lowerToKey = Object.keys(obj).reduce((acc, key) => {
  acc[key.toLowerCase()] = key;
}, {});
...
const realKey = lowerToKey[pathArray[i].toLowerCase()];
```

File: libraries/botbuilder-dialogs/src/memory/scopes/settingsMemoryScope.ts [hunk 2]
```diff
@@ -205,4 +221,29 @@ export class SettingsMemoryScope extends MemoryScope {
...
     }
+
+    private static filterSettings(settings: Record<string, unknown>): Record<string, unknown> {
+        const result = Object.assign({}, settings);
+        SettingsMemoryScope.blockingList.forEach((path) => SettingsMemoryScope.deletePropertyPath(result, path));
+        return result;
+    }
+
+    private static deletePropertyPath(obj, path: string): void {
+        if (!obj || !path) {
```
Comment (joshgummersall) [inline libraries/botbuilder-dialogs/src/memory/scopes/settingsMemoryScope.ts:232]: Perhaps `if (!obj || !path?.length)`

**Review by joshgummersall — COMMENTED** [review]
Couple quick suggestions, otherwise looks good.

**Comment by joshgummersall** [comment]
@Danieladu I made the quick fixes I suggested so we're good to go.

**Review by joshgummersall — DISMISSED** [review]

File: libraries/botbuilder-dialogs-adaptive-testing/tests/resources/SettingsStateTests/SettingsStateTests_SettingsTest.test.dialog [hunk 3]
```diff
@@ -10,14 +10,6 @@
...
                         "$kind": "Microsoft.SendActivity",
                         "activity": "${settings.MicrosoftAppId}"
-                    },
-                    {
-                        "$kind": "Microsoft.SendActivity",
-                        "activity": "${settings.MicrosoftAppPassword}"
-                    },
-                    {
-                        "$kind": "Microsoft.SendActivity",
-                        "activity": "${settings.ApplicationInsightsInstrumentationKey}"
```
Comment (mdrichardson) [inline libraries/botbuilder-dialogs-adaptive-testing/tests/resources/SettingsStateTests/SettingsStateTests_SettingsTest.test.dialog:20]: In [.NET, this test was adjusted to ensure these settings are blocked](https://github.com/microsoft/botbuilder-dotnet/pull/5641/files#diff-f1ca39c30c78da07bd745b2ac47e01e2c33b5cb02e96c49aaeec0c476c1ca2acR21). Can we do the same thing, here?

Comment (boydc2014) [reply libraries/botbuilder-dialogs-adaptive-testing/tests/resources/SettingsStateTests/SettingsStateTests_SettingsTest.test.dialog]: Added by the latest commit

**Review by joshgummersall — APPROVED** [review]

**Pull request merged**